test/unit/assorted/server_discovery_and_monitoring.spec.test.ts

```diff
@@ -0,0 +1,513 @@
+          relaxed: true
+        });
+
+        if (!isRecord(result) || Array.isArray(result)) {
+          throw new Error(`${filePath} did not contain a top-level object`);
+        }
+
+        result.type = testType;
+        result.fileName = path.join(testType, f); // unused but helpful when debugging
+        return result;
+      });
+  }
+
+  return tests;
+}
+
+interface SDAMTest {
+  description: string;
+  uri: string;
+  phases: SDAMPhase[];
+}
+/**
+ * A phase of the test optionally sends inputs to the client,
+ * then tests the client's resulting TopologyDescription.
+ */
+type SDAMPhase =
+  | {
+      description?: string;
+      applicationErrors: ApplicationError[];
+      outcome: TopologyDescriptionOutcome;
+    }
+  | {
+      description?: string;
+      responses?: SDAMResponse[];
+      outcome: MonitoringOutcome | TopologyDescriptionOutcome;
+    };
+
+interface MonitoringOutcome {
+  events: typeof SDAM_EVENT_CLASSES[keyof typeof SDAM_EVENT_CLASSES][];
+}
+interface OutcomeServerDescription {
+  type?: string;
+  setName?: string;
+  setVersion?: number;
+  electionId?: ObjectId | null;
+  logicalSessionTimeoutMinutes?: number;
+  minWireVersion?: number;
+  maxWireVersion?: number;
+  topologyVersion?: TopologyVersion;
+  pool?: { generation: number };
+}
+interface TopologyDescriptionOutcome {
+  topologyType: string;
+  setName?: string;
+  servers?: Record<string, OutcomeServerDescription>;
+  logicalSessionTimeoutMinutes?: number;
+  maxSetVersion?: number;
+  maxElectionId?: ObjectId;
+  compatible: false | undefined;
+}
+type SDAMResponse = [serverAddress: string, hello: Document];
+type ApplicationErrorCommon = {
+  address: string;
+  generation?: number;
+  maxWireVersion?: number;
+  when: 'beforeHandshakeCompletes' | 'afterHandshakeCompletes';
+};
+type ApplicationError =
+  | (ApplicationErrorCommon & { type: 'network' | 'timeout' })
+  | (ApplicationErrorCommon & { type: 'command'; response: Document });
+
+function isTopologyDescriptionOutcome(outcome: any): outcome is TopologyDescriptionOutcome {
+  try {
+    assertTopologyDescriptionOutcome(outcome);
+    return true;
+  } catch {
+    return false;
+  }
+}
+
+function assertTopologyDescriptionOutcome(
+  outcome: any
+): asserts outcome is TopologyDescriptionOutcome {
+  expect(outcome).to.be.an('object').that.is.not.null;
+  expect(outcome).to.have.property('topologyType').that.is.a('string');
+  expect(outcome).to.have.property('servers').that.is.an('object');
+  // The type annotation helps keep this sync-ed with the typescript interface
+  const knownTopologyDescriptionOutcomeKeys: ReadonlyArray<keyof TopologyDescriptionOutcome> = [
+    'topologyType',
+    'setName',
+    'servers',
+    'logicalSessionTimeoutMinutes',
+    'maxSetVersion',
+    'maxElectionId',
+    'compatible'
+  ] as const;
+  expect(outcome).to.have.any.keys(knownTopologyDescriptionOutcomeKeys);
```

Comment:
if you're trying to assert that the outcome keys are a subset of the known keys, you want to do:
```js
for (const key of Object.keys(outcome)) {
  expect(knownTopologyDescriptionOutcomeKeys).to.include(key);
}
```